src/wrapper/java_vm/vm.rs

```diff
@@ -49,45 +52,88 @@ impl JavaVM {
         self.0
     }
 
+    /// Attaches the current thread to a JVM. Calling this for a thread that is already attached
+    /// is a no-op.
+    ///
+    /// Thread will be detached automatically when finished.
+    ///
+    /// Attached threads [block JVM exit][block].
+    ///
+    /// [block]: https://docs.oracle.com/en/java/javase/12/docs/specs/jni/invocation.html#unloading-the-vm
+    pub fn attach_current_thread_permanently(&self) -> Result<JNIEnv> {
+        match self.get_env() {
+            Ok(env) => Ok(env),
+            Err(_) => {
+                unsafe {
+                    let env_ptr = InternalAttachGuard::create_and_attach(
+                        self.get_java_vm_pointer(),
+                        ThreadType::Normal
+                    )?;
+                    JNIEnv::from_raw(env_ptr)
+                }
+            }
+        }
+    }
+
     /// Attaches the current thread to a Java VM. The resulting `AttachGuard`
     /// can be dereferenced to a `JNIEnv` and automatically detaches the thread
     /// when dropped. Calling this for a thread that is already attached is a no-op.
+    ///
+    /// Attached threads [block JVM exit][block].
+    ///
+    /// Attaching and detaching is time-consuming operation, therefore multiple short-term attach
+    /// operations on the same thread should be avoided or replaced with
+    /// `attach_current_thread_permanently`.
+    ///
+    /// [block]: https://docs.oracle.com/en/java/javase/12/docs/specs/jni/invocation.html#unloading-the-vm
     pub fn attach_current_thread(&self) -> Result<AttachGuard> {
-        let (env, requires_detach) = match self.get_env() {
-            Ok(env) => (env, false),
-            Err(_) => unsafe {
-                let mut ptr = ptr::null_mut();
-                let res = java_vm_unchecked!(self.0, AttachCurrentThread, &mut ptr, ptr::null_mut());
-                jni_error_code_to_result(res)?;
-
-                let env = JNIEnv::from_raw(ptr as *mut sys::JNIEnv)?;
-                (env, true)
+        match self.get_env() {
+            Ok(env) => {
+                Ok(AttachGuard::new_nested(env))
             },
-        };
+            Err(_) => {
+                let env_ptr = InternalAttachGuard::create_and_attach(
+                    self.get_java_vm_pointer(),
+                    ThreadType::Normal
+                )?;
+                let env = unsafe {
+                    JNIEnv::from_raw(env_ptr)?
+                };
```

Comment:
I see a very similar code in `attach_current_thread_permanently`, `attach_current_thread` and `attach_current_thread_as_daemon`. Can it be moved into a function?

Reply:
Decided to move `create_and_attach` into JavaVM as internal method - it reduces the code duplication significantly, but maybe brings some mess. What do you think?